synapse/state/__init__.py

```diff
@@ -120,18 +123,46 @@ async def get_state(
         if self._state is not None:
             return self._state
 
-        assert self.state_group is not None
+        if self.state_group is not None:
+            return await state_storage.get_state_ids_for_group(
+                self.state_group, state_filter
+            )
+
+        assert self.prev_group is not None and self.delta_ids is not None
 
-        return await state_storage.get_state_ids_for_group(
-            self.state_group, state_filter
+        prev_state = await state_storage.get_state_ids_for_group(
+            self.prev_group, state_filter
         )
 
+        # ChainMap expects MutableMapping, but since we're using it immutably
+        # its safe to give it immutable maps.
+        return ChainMap(self.delta_ids, prev_state)  # type: ignore[arg-type]
+
+    def set_state_group(self, state_group: int) -> None:
+        """Update the state group assigned to this state (e.g. after we've
+        persisted it)
+        """
+
+        self.state_group = state_group
+
+        # We clear out the state as we know longer need to explicitly keep it in
+        # the `state_cache` (as the store state group cache will do that.)
+        self._state = None
+
     def __len__(self) -> int:
         # The len should is used to estimate how large this cache entry is, for
-        # cache eviction purposes. This is why if `self.state` is None it's fine
-        # to return 1.
+        # cache eviction purposes. This is why it's fine to return 1 if we're
+        # not storing any state.
```

Comment:
I know this is not a new thing introduced by this PR, but why do we return 1 if we're not storing any state (as opposed to 0)?

Reply:
It's an estimate of the memory usage of the cached value, basically. Even if we don't store any state the `_StateCacheEntry` still takes a bit of memory.

If we returned 0 then empty entries wouldn't change the size of the cache, so you could potentially have an unbounded number of items in the cache.

Reply:
Ah that makes sense, thanks for explaining!